src/team_lib/mhba/xccl_mhba_mkeys.c

```diff
@@ -0,0 +1,283 @@
+    node->recv_mkey_entries = (struct mlx5dv_mr_interleaved *) calloc(node->sbgp->group_size,
+                                                                      sizeof(struct mlx5dv_mr_interleaved));
+    return XCCL_OK;
+}
+
+xccl_status_t poll_umr_cq(xccl_mhba_context_t *ctx) {
+
+    int expected_completions_num = 1;
+    struct ibv_wc work_completion[1];
+    memset(&work_completion, 0, expected_completions_num * sizeof(struct ibv_wc));
+    int num_completions;
+    int return_code = poll_cq(ctx->umr_cq, expected_completions_num, &num_completions, work_completion);
+    if (return_code <= 0) {
+        xccl_mhba_error("UMR WQE on UMR QP failed (error %d)", return_code);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    if (num_completions != expected_completions_num) {
+        xccl_mhba_error("UMR CQ didn't return the expected number of completions (expected %d, returned %d)",
+                        expected_completions_num, num_completions);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    if (work_completion[0].status != IBV_WC_SUCCESS) {
+        xccl_mhba_error("UMR CQ returned completion with status %s (%d)",
+                        ibv_wc_status_str(work_completion[0].status), work_completion[0].status);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    if (work_completion[0].opcode != MLX5DV_WC_UMR) {
+        xccl_mhba_error("Got unexpected completion opcode for the UMR QP: %d", work_completion[0].opcode);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    xccl_mhba_debug("Successfully executed the UMR WQE");
+    return XCCL_OK;
+}
+
+// Execute the UMR WQE for populating the UMR's MasterMKey
+xccl_status_t
+populate_mkey(xccl_mhba_context_t *ctx, xccl_mhba_node_t *node, int mem_access_flags, struct mlx5dv_mkey *mkey,
+              struct mlx5dv_mr_interleaved *mkey_entries) {
+    xccl_status_t status;
+    xccl_mhba_debug("Execute the UMR WQE for populating the MasterMKey");
+    ibv_wr_start(ctx->umr_qpx);
+    ctx->umr_qpx->wr_id = 1; // First (and only) WR todo: check meaning
+    int repeat_count = node->sbgp->group_size / node->block_size; // todo check calc
+    mlx5dv_wr_mr_interleaved(ctx->umr_mlx5dv_qp_ex, mkey, mem_access_flags,
+                             repeat_count, node->sbgp->group_size,
+                             mkey_entries);
+    status = poll_umr_cq(ctx);
+    if (status != XCCL_OK) {
+        return status;
+    }
+    return XCCL_OK;
+}
+
+xccl_status_t
+update_mkey_entries(xccl_mhba_context_t *ctx, xccl_mhba_node_t *node, struct mlx5dv_mkey *mkey,
+                    struct mlx5dv_mr_interleaved *mkey_entries,
+                    void *buffer_start, int mem_access_flags) {
+    xccl_status_t status;
+    xccl_mhba_debug("Updating Strided-KLM entries for the recv MasterMKey");
+    int i;
+    for (i = 0; i < node->sbgp->group_size; i++) {
+        struct ibv_mr *curr_bf = (struct ibv_mr *) (buffer_start + i * node->size_of_data_unit);
```

Comment:
Why do we need this "extra" copy. Currently, node DR ranks register memory and then save ibv_mr to shared memory storage. Why not allcoating the shared memory storage with MHBA_DATA_SIZE sizeof(struct mlx5dv_mr_interleaved) and then each rank will populate its own mlx5dv_mr_interleaved struct in the shared memory. The AST then can use the shared memory storage directly as array of mkey_entires. and you don't need 2 more arrays for mkey_entries on ASR rank.?

Reply:
Agree. I'll change